File: src/SendereCommons/RemoteUser.java
package SendereCommons;

import com.google.protobuf.Any;
import lombok.Data;
import lombok.Getter;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.net.SocketException;
import java.security.InvalidKeyException;
import java.security.KeyFactory;
import java.security.NoSuchAlgorithmException;
import java.security.PublicKey;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.X509EncodedKeySpec;
import java.util.Objects;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;

public abstract class RemoteUser {

    private String nickname;
    private long hash;
    private Socket socket;
    private int port;
    private boolean identified = false;
    @Getter
    private InputStream inputStream;
    @Getter
    private OutputStream outputStream;
    @Getter
    private boolean disconnected = false;
    private long lastAliveTime;
    @Getter
    private int authenticationStage = 0;
    private SimpleNetworkInterface networkInterface;
    private BlockingQueue<SendingQueueElement> sendingQueue = new ArrayBlockingQueue<SendingQueueElement>(10);
    private PublicKey rsaPublicKey;
    private long secret = 0;
    private byte encryptedSecret[];


    public RemoteUser(String nickname, long hash, Socket socket) throws IOException {
        identify(nickname, hash);
        commonInitialization(socket);
    }

    public RemoteUser(Socket socket) throws IOException {
        commonInitialization(socket);
        Timer identifyTimer = new Timer();
        identifyTimer.schedule(new TimerTask() {
            @Override
            public void run() {
                if (!identified)
                    onDisconnectInternal();
            }
        }, 5000);
    }

    public int getPort() {
        return socket.getPort();
    }

    private void commonInitialization(Socket socket) throws IOException {
        this.socket = socket;
        inputStream = socket.getInputStream();
        outputStream = socket.getOutputStream();
        lastAliveTime = System.currentTimeMillis();
        doReceiving();
        doSending();
        doAlive();
    }

    private void doSending() {
//        final Thread sendingThread = new Thread(new Runnable() {
//            @Override
//            public void run() {
//                while (true) {
//                    try {
//                        SendingQueueElement element = sendingQueue.take();
//                        int length = element.getLength();
//                        outputStream.write(new byte[]{(byte) ((length & 0x00FF0000) >> 16), (byte) ((length & 0x0000FF00) >> 8), (byte) (length & 0x000000FF)});
//                        outputStream.write(element.flags);
//                        if (length == 0)
//                            continue;
//                        for (byte[] dataPiece : element.getDatas())
//                            outputStream.write(dataPiece);
//                    } catch (IOException e) {
//                        e.printStackTrace();
//                        onDisconnectInternal();
//                    } catch (InterruptedException e) {
//                        e.printStackTrace();
//                    }
//                }
//            }
//        });
//        sendingThread.start();
    }

    private void doAlive() {
        Timer timer = new Timer();
        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                sendAlive();
            }
        }, 0, 2000);
    }

    private void doReceiving() {
        Thread receiverThread = new Thread(new Runnable() {
            @Override
            public void run() {
                while (!disconnected) {
                    int read;
                    byte[] packetLength = new byte[4];
                    byte flags;
                    int length;
                    try {
//                        while (inputStream.available() < 4 && !disconnected) {
////                            if (System.currentTimeMillis() > lastAliveTime + 5000)
////                                RemoteUser.this.onDisconnectInternal();
//                            System.out.println("waiting...");
//                            Thread.sleep(2000);
//                        }
                        int headerRead = 0;
                        while (headerRead<4 && !disconnected){
                            headerRead+=inputStream.read(packetLength,headerRead,4-headerRead);
                        }
                        length = ((packetLength[0] + (packetLength[0] >= 0 ? 0 : 256)) << 16) + ((packetLength[1] + (packetLength[1] >= 0 ? 0 : 256)) << 8) + packetLength[2] + (packetLength[2] >= 0 ? 0 : 256);
                        //47 is '/' symbol's code
                        //18.09.2019
//                    if(packetLength[3]!=47)
//                        continue;
                       // flags = (byte) inputStream.read();
                        read = 0;
                        byte[] data = new byte[length];
                        while (read < length && !disconnected)
                            read += inputStream.read(data, read, length - read);
                        lastAliveTime = System.currentTimeMillis();
                        RemoteUser.this.onReceive(packetLength[3], data);

                    } catch (SocketException e) {
                        RemoteUser.this.onDisconnectInternal();
                    } catch (IOException e) {
                        //e.printStackTrace();
                    }
                }
                try {
                    inputStream.close();
                    outputStream.close();
                } catch (IOException e) {
                    //Okay
                    //10.09.2019
                    e.printStackTrace();
                }
            }
        });
        receiverThread.start();
    }


    private void onDisconnectInternal() {
        if (disconnected)
            return;
        disconnected = true;
        if (isIdentified())
            onDisconnect();
    }

    protected abstract void onDisconnect();

    boolean sendAlive() {
        // Send headless packet to keep connection
        // 11.04.2021 huku

//        if (disconnected)
//            return false;
//        try {
//            // TODO: 11.04.2021 send flags: stub
//            sendingQueue.put(new SendingQueueElement(0, (byte) 0, null));
//            return true;
//        } catch (InterruptedException e) {
//            e.printStackTrace();
//        }

        sendMessage((byte) 0);
        return false;
    }

    /**
     * @param header
     * @param data
     * @param length
     * @return
     * @deprecated go use partial send
     */
    @Deprecated
    boolean sendMessage(byte header, byte[] data, int length) {
        return sendMessage((byte) 0, new byte[]{header}, data);
    }


    public boolean sendMessage(byte flags, byte[]... data) {
        if (disconnected)
            return false;
        int length = 0;
        for (byte[] dataPiece : data) {
            length += dataPiece.length;
        }
        //byte[] byteLength = new byte[]{(byte) ((length & 0x00FF0000) >> 16), (byte) ((length & 0x0000FF00) >> 8), (byte) (length & 0x000000FF)};
//        try {
//            //Don't pay attention on "Synchronization on a non-final field" warning
//            //output stream never changing even if javac can't understand it
//            //19.09.2019
//            sendingQueue.put(new SendingQueueElement(length, flags, data));
//            return true;
//        } catch (InterruptedException e) {
//            e.printStackTrace();
//        }
        synchronized (outputStream) {
            try {
                outputStream.write(new byte[]{(byte) ((length & 0x00FF0000) >> 16), (byte) ((length & 0x0000FF00) >> 8), (byte) (length & 0x000000FF)});
                outputStream.write(flags);
                if (length == 0)
                    return true;
                for (byte[] dataPiece : data)
                    outputStream.write(dataPiece);
            } catch (IOException e) {
                e.printStackTrace();
                onDisconnectInternal();
            }
        }
        return false;
    }

    @Override
    public String toString() {
        return nickname;
    }

    public String getNickname() {
        return nickname;
    }

    public long getHash() {
        return hash;
    }

    void identify(String nickname, long hash) {
        this.nickname = nickname;
        this.hash = hash;
        identified = true;
    }

    private boolean isIdentified() {
        return identified;
    }

    public abstract void onReceive(byte flags, byte[] data);

    public String getAddress() {
        return socket.getInetAddress().getHostAddress();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        RemoteUser that = (RemoteUser) o;
        return hash == that.hash && port == that.port && identified == that.identified && Objects.equals(nickname, that.nickname) && Objects.equals(socket, that.socket);
    }

    @Override
    public int hashCode() {
        return Objects.hash(nickname, hash);
    }

    @Data
    private static class SendingQueueElement {
        private int length;
        private byte flags;
        private byte[][] datas;

        public SendingQueueElement(int length, byte flags, byte[][] datas) {
            this.length = length;
            this.flags = flags;
            this.datas = datas;
        }
    }

    public boolean doAuthenticationStage1(long secret, byte[] encryptedSecret, byte[] publicKeyBytes){
        try {
            rsaPublicKey = KeyFactory.getInstance("RSA").generatePublic(new X509EncodedKeySpec(publicKeyBytes));
            this.secret = secret;
            long decryptedSecret = Sendere.decrypt(encryptedSecret, rsaPublicKey);
            if (decryptedSecret == secret){
                this.encryptedSecret = encryptedSecret;
                authenticationStage=1;
                return true;
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }
}

File: src/sendereCommons/RemoteUser.kt
package sendereCommons

import lombok.Data
import lombok.Getter
import lombok.Setter
import lombok.SneakyThrows
import sendereCommons.Sendere.Companion.decrypt
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.Socket
import java.net.SocketException
import java.security.KeyFactory
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.PublicKey
import java.security.spec.X509EncodedKeySpec
import java.util.*
import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.BlockingQueue
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.interfaces.DHPublicKey
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

abstract class RemoteUser {
    var nickname: String? = null
    var hash: Long = 0
    private val port = 0
    private var isIdentified = false

    private var disconnected = false
    private var lastAliveTime: Long = 0

    private lateinit var inputStream: InputStream
    private lateinit var outputStream: OutputStream
    private lateinit var socket: Socket

    lateinit var keyPair: KeyPair
    private lateinit var keyAgreement: KeyAgreement
    private lateinit var sharedSecret: ByteArray
    private lateinit var encryptionCipher: Cipher
    private lateinit var decryptionCipher: Cipher

    var encryptionDone = false
    fun createKeyAgreement(): Boolean {
        return try {
            val aliceKpairGen = KeyPairGenerator.getInstance("DH")
            aliceKpairGen.initialize(2048)
            keyPair = aliceKpairGen.generateKeyPair()
            keyAgreement = KeyAgreement.getInstance("DH")
            keyAgreement.init(keyPair.private)
            println("Stage 1 done")
            true
        } catch (e: Exception) {
            // algorithm not supported
            println("Stage 1 failed")
            false
        }
    }

    fun createKeyAgreementFromReceivedKey(publicKey: DHPublicKey, ivBytes: ByteArray?): Boolean {
        return try {
            val bobKpairGen = KeyPairGenerator.getInstance("DH")
            bobKpairGen.initialize(publicKey.params)
            keyPair = bobKpairGen.generateKeyPair()
            keyAgreement = KeyAgreement.getInstance("DH")
            keyAgreement.init(keyPair.private)
            keyAgreement.doPhase(publicKey, true)
            sharedSecret = keyAgreement.generateSecret()
            val bobAesKey = SecretKeySpec(sharedSecret, 0, 16, "AES")
            encryptionCipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
            encryptionCipher.init(Cipher.ENCRYPT_MODE, bobAesKey, IvParameterSpec(ivBytes))
            decryptionCipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
            decryptionCipher.init(Cipher.DECRYPT_MODE, bobAesKey, IvParameterSpec(ivBytes))
            println("Stage 2 done")
            true
        } catch (e: Exception) {
            // algorithm not supported
            e.printStackTrace()
            println("Stage 2 failed")
            false
        }
    }

    @SneakyThrows
    fun encrypt(cleartext: ByteArray): ByteArray {
        return encryptionCipher.doFinal(cleartext)
    }

    @SneakyThrows
    fun decrypt(encrypted: ByteArray): ByteArray {
        return decryptionCipher.doFinal(encrypted)
    }

    fun updateKeyAgreementWithReceivedKey(publicKey: DHPublicKey?, ivBytes: ByteArray): Boolean {
        return try {
            keyAgreement.doPhase(publicKey, true)
            sharedSecret = keyAgreement.generateSecret()
            val aliceAesKey = SecretKeySpec(sharedSecret, 0, 16, "AES")
            encryptionCipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
            encryptionCipher.init(Cipher.ENCRYPT_MODE, aliceAesKey, IvParameterSpec(ivBytes))
            decryptionCipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
            decryptionCipher.init(Cipher.DECRYPT_MODE, aliceAesKey, IvParameterSpec(ivBytes))
            println("Stage 3 done")
            true
        } catch (e: Exception) {
            println("Stage 3 failed")
            false
        }
    }

    constructor(nickname: String?, hash: Long, socket: Socket) {
        identify(nickname, hash)
        commonInitialization(socket)
    }

    constructor(socket: Socket) {
        commonInitialization(socket)
        val identifyTimer = Timer()
        identifyTimer.schedule(object : TimerTask() {
            override fun run() {
                if (!isIdentified) onDisconnectInternal()
            }
        }, 5000)
    }

    fun getPort(): Int {
        return socket.port
    }

    @Throws(IOException::class)
    private fun commonInitialization(socket: Socket) {
        this.socket = socket
        inputStream = socket.getInputStream()
        outputStream = socket.getOutputStream()
        lastAliveTime = System.currentTimeMillis()
        doReceiving()
        doAlive()
    }

    private fun doAlive() {
        val timer = Timer()
        timer.schedule(object : TimerTask() {
            override fun run() {
                sendAlive()
            }
        }, 0, 2000)
    }

    private fun doReceiving() {
        val receiverThread = Thread {
            while (!disconnected) {
                var read: Int
                val packetLength = ByteArray(4)
                var flags: Byte
                var length: Int
                try {
//                        while (inputStream.available() < 4 && !disconnected) {
////                            if (System.currentTimeMillis() > lastAliveTime + 5000)
////                                RemoteUser.this.onDisconnectInternal();
//                            System.out.println("waiting...");
//                            Thread.sleep(2000);
//                        }
                    var headerRead = 0
                    while (headerRead < 4 && !disconnected) {
                        headerRead += inputStream.read(packetLength, headerRead, 4 - headerRead)
                    }
                    length = (packetLength[0].toUByte().toInt() shl 16) +
                            + (packetLength[1].toUByte().toInt() shl 8) +
                            + (packetLength[2].toUByte().toInt())
                   // length = (packetLength[0] + if (packetLength[0] >= 0) 0 else 256 shl 16) + (packetLength[1] + if (packetLength[1] >= 0) 0 else 256 shl 8) + packetLength[2] + if (packetLength[2] >= 0) 0 else 256
                    //47 is '/' symbol's code
                    //18.09.2019
//                    if(packetLength[3]!=47)
//                        continue;
                    // flags = (byte) inputStream.read();
                    read = 0
                    val data = ByteArray(length)
                    while (read < length && !disconnected) read += inputStream.read(data, read, length - read)
                    lastAliveTime = System.currentTimeMillis()
                    onReceived(packetLength[3], data)
                } catch (e: SocketException) {
                    onDisconnectInternal()
                } catch (e: IOException) {
                    //e.printStackTrace();
                }
            }
            try {
                inputStream.close()
                outputStream.close()
            } catch (e: IOException) {
                //Okay
                //10.09.2019
                e.printStackTrace()
            }
        }
        receiverThread.start()
    }

    private fun onDisconnectInternal() {
        if (disconnected) return
        disconnected = true
        if (isIdentified) onDisconnected()
    }

    protected abstract fun onDisconnected()
    fun sendAlive(): Boolean {
        // Send headless packet to keep connection
        // 11.04.2021 huku

//        if (disconnected)
//            return false;
//        try {
//            // TODO: 11.04.2021 send flags: stub
//            sendingQueue.put(new SendingQueueElement(0, (byte) 0, null));
//            return true;
//        } catch (InterruptedException e) {
//            e.printStackTrace();
//        }
        sendMessage(0.toByte(), byteArrayOf())
        return false
    }

    fun sendMessage(flags: Byte, data: ByteArray): Boolean {
        if (disconnected) return false
        var length = data.size
        // Don't pay attention on "Synchronization on a non-final field" warning
        // output stream never changing even if javac can't understand it
        // 19.09.2019 huku
        synchronized(outputStream) {
            try {
                outputStream.write(byteArrayOf((length and 0x00FF0000 shr 16).toByte(), (length and 0x0000FF00 shr 8).toByte(), (length and 0x000000FF).toByte()))
                outputStream.write(flags.toInt())
                if (length == 0) return true
                outputStream.write(data)
                return true
            } catch (e: IOException) {
                onDisconnectInternal()
            }
        }
        return false
    }

    override fun toString(): String {
        return nickname!!
    }

    fun identify(nickname: String?, hash: Long) {
        this.nickname = nickname
        this.hash = hash
        isIdentified = true
    }

    abstract fun onReceived(flags: Byte, data: ByteArray)
    val address: String
        get() = socket.inetAddress.hostAddress

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val that = o as RemoteUser
        return hash == that.hash && port == that.port && isIdentified == that.isIdentified && nickname == that.nickname && socket == that.socket
    }

    override fun hashCode(): Int {
        return Objects.hash(nickname, hash)
    }

    @Data
    private class SendingQueueElement(private val length: Int, private val flags: Byte, private val datas: Array<ByteArray>)
}